### ISSUE TITLE
Running rules that require type-check

### ISSUE THREAD

**Issue by steverob2k** [Microsoft/vscode-tslint#70]
Thanks for the work on this plugin it is extremely useful.

tslint version: 3.14.0

I have updated my tslint rules to include 2 new ones 

```
"restrict-plus-operands": true 
"no-for-in-array": true,
```

Running via the tslint command line 
`tslint --type-check --project tsconfig.json "./app/**/*.ts"`

I get a list of files where the rules have failed. However running them through the vscode extension doesn't produce any warnings.

From the tslint documentation there are additional steps that are required to run rules which require type check. 

Is this likely to be supported anytime soon?

**Comment by glen-84** [comment]
Maybe two new extension options could be added:

`tslint.project` and `tslint.typeCheck`.

Another option would be an `options` object:

```json
"tslint.options": {
    "project": "tsconfig.json",
    "typeCheck": true
}
```

**Comment by egamma** [comment]
@glen-84 I agree that these extensions should be exposed as well. 

In general we favor 'flat' settings so `tslint.project` and `tslint.typeCheck` are my preference.

**Comment by egamma** [comment]
Actually this is more involved than adding some options that you pass on the tslint API. 

Rules that require type checking assume that all files of the project are linted during a a validation https://github.com/palantir/tslint#type-checking. 

Linting all files in a project on each change is expensive and makes the linting very slow unless the linting is done in an incremental way.  As far as I know tslint doesn't provide support to make the validation incremental.

**Comment by skysteve** [comment]
Any update on this please given it's been a few months since this issue was opened?

Thanks
Steve

**Comment by egamma** [comment]
@skysteve I'm sorry, but currently there are no plans to add support for typechecked rules on each change (save/type). It doesn't make sense as explained here https://github.com/Microsoft/vscode-tslint/issues/70#issuecomment-241041929.  

My recommendation is to run tslint with typecheck as a task as is documented in the README in the section `Using the extension with tasks running tslint`. Then you get squiggles and warnings in the Problems panel.

**Comment by aluanhaddad** [comment]
> Given this it doesn't makes sense to me to support typechecked rules on each change. What I could imagine is a separate command that the user can invoke manually to also run tslint with typechecked rules.

@egama That makes perfect sense. However given that, would it not also make sense to for the plugin, if at all possible, to simply ignore these rules instead as failing as it currently does.
Specifically it is awkward to do as you recommend

> My recommendation is to run tslint with typecheck as a task as is documented in the README in the section Using the extension with tasks running tslint. Then you get squiggles and warnings in the Problems panel.

because it requires having multiple tslint configuration files even for small projects.

PS: (Thanks for adding the status indicator for the plugin it is extremely helpful!)

**Comment by egamma** [comment]
@aluanhaddad I see the pain point and agree that this is unhappy. However, it looks like tslint currently provides no way to configure it, so does not run rules that require type checking when type checking is not enabled. 

Also there is no easy way to distinguish the different exceptions from tslint other than relying on string matching on the string suffix `requires type checking` (see below).

My preferred option is that tslint provides an option `no-typecheckrules` at the API level or that tslint adds some error code to exceptions. 

![image](https://cloud.githubusercontent.com/assets/172399/20648395/cfa040b2-b4a5-11e6-8388-40b93c9a1da5.png)

**Comment by egamma** [comment]
I've tried to implement the exception filtering. The issue is that tslint bails out completely when there is such an exception, that is, it doesn't validate with any of the other rules.

So the best recommendation is to have 2 tslint.json files and separate the rules with type checking from into a separate tslint.json and then use `extends` to include the non type checking rules.

**Comment by aluanhaddad** [comment]
@egamma thank you for looking into this and for trying to implement it. It's very unfortunate that the API is structured like that. I'm going to open an issue on  https://github.com/palantir/tslint in the meantime I'll use your suggestion.

**Comment by JKillian** [comment]
> I've tried to implement the exception filtering. The issue is that tslint bails out completely when there is such an exception, that is, it doesn't validate with any of the other rules.

This behavior was changed in TSLint recently (https://github.com/palantir/tslint/issues/2169) so that TSLint now just prints a warning message and continues linting. That should remove one of the big issues mentioned above, so it ought to be possible to just use one `tslint.json` file now instead of multiple.

**Comment by aluanhaddad** [comment]
Thanks @JKillian!

**Comment by egamma** [comment]
This is good news, thanks. 

Also once we can run tslint as a TypeScript server extension (see #171), then rules that require type check will be supported.

**Comment by aluanhaddad** [comment]
@egamma that's awesome. Really looking forward to that!

**Comment by egamma** [comment]
A first version of the tslint-language-server extension is now available pls see A version that works in VS Code and Eclipse with tslint 4/5 is now available pls see https://github.com/angelozerr/tslint-language-service/blob/master/README.md.

I'm closing this issue, the right way to support this feature request is the tslint-language-server plugin.

**Comment by egamma** [comment]
A first version of the tslint-language-server extension is now available pls see A version that works in VS Code and Eclipse with tslint 4/5 is now available pls see https://github.com/angelozerr/tslint-language-service/blob/master/README.md.

I'm closing this issue, the right way to support this feature request is the tslint-language-server plugin.

**Comment by aleksey-bykov** [comment]
sorry for ruining the party but the pluging infrastructure is specifically NOT designed for such kind of use cases:

> Examples of things language plugins cannot do:
> - Customize the type system to change what is or isn't an error when running tsc

saurce: https://github.com/Microsoft/TypeScript/wiki/Writing-a-Language-Service-Plugin#whats-a-language-service-plugin

**Comment by egamma** [comment]
>sorry for ruining the party but the pluging infrastructure is specifically NOT designed for such kind of use cases

@aleksey-bykov I didn't come to the same conclusion, the tslint plugin doesn't customize the type system or change the diagnostics. The tslint plugin **augments** the existing diagnostics provided by TypeScript and **augments** the available code actions/quick fixes.

**Comment by aleksey-bykov** [comment]
there is no way to get all linting errors though a plugin, say at running the build, is there?

**Comment by KeitIG** [comment]
For those struggling, as a workaround you can use

```
compilerOptions: {
  noUnusedLocals: true
}
```

in your `tsconfig.json`

**Comment by egamma** [comment]
@KeitIG thanks. It is also documented in the FAQ in the README https://github.com/Microsoft/vscode-tslint/blob/master/tslint/README.md#faq.
